src/jit/lsraxarch.cpp

```diff
@@ -2362,6 +2362,14 @@ void LinearScan::BuildHWIntrinsic(GenTreeHWIntrinsic* intrinsicTree)
             useList.Last()->info.isTgtPref = true;
             break;
 
+        case NI_SSE2_MaskMove:
+        {
+            // SSE2 MaskMove hardcodes the destination (op3) in DI/EDI/RDI
+            LocationInfoListNode* op3Info = useList.Begin()->Next()->Next();
+            op3Info->info.setSrcCandidates(this, RBM_EDI);
```

Comment:
@CarolEidt, what is the proper way to do this? The LSRA doesn't seem to respect this (or the similar pattern used by `NI_SSE41_BlendVariable` for `XMM0` below)...

Reply:
The codgen should insert register copy if `op3Reg` is not `EDI`.

Reply:
@fiigii, I don't think that is quite correct (and would like confirmation if it is).

I would think that the register allocator needs to explicitly give us the required register, otherwise we would be consuming it without its knowledge and potentially overwriting whatever value the JIT thinks is in there.

Reply:
Looks like this comment on `checkConflictingDefUse` is relevant: https://github.com/dotnet/coreclr/blob/master/src/jit/lsrabuild.cpp#L365

It would be helpful if there was a "hack" we could do for this to enforce that the required register is actually used.

Reply:
Man, this was discussed to death in the past. The RA will allocate the required register for you but you have to move the value to the required register if it's not already there. There's little reason for the RA to automatically insert a copy node for this.

Reply:
@mikedn, how can I check (in codegen) that the RA has actually allocated the required register?

Based on the JitDump and other information available, I can't see anywhere that RDI is actually "reserved" for my use.

Reply:
@mikedn is correct. If the specified requirement is fixed (i.e. not a set of registers), then the register allocator will ensure that it is availalble, but it is the responsibility of the code generator to check and do the move if it isn't in the expected register.

> It would be helpful if there was a "hack" we could do for this to enforce that the required register is actually used.

The reason this isn't done is that the register allocator sometimes can't (or won't, for other reasons) allocate that register at the definition. If it had to ensure that the register was actually used, it would have to insert a move, which is extra work and extra generated IR. It will indeed insert a move if the requirement is a subset of registers, since then the code generator won't know what register was made available.

This comment you cite is related but not directly. This addresses the issue that it might have been possible for the register allocator to reserve that register earlier, but it was already occupied by a definition that didn't "see" the fixed requirement as occuring at the def (but only at the use). Moving the fixed requirement to the def can be done (and has been done experimentally), but requires more work in the register allocator (because the fixed RefPosition is created when we haven't yet seen the use) and results in some cases of over-constraint that actually make things worse. That's want the "require further analysis" is about.

Reply:
If you look at the allocation table in the jitdump, you should be able to see that RDI isn't occupied at the time of the use. You can search for "final allocation" to see what the register allocator thinks is in each register at each virtual cycle (LsraLocation).

Reply:
@CarolEidt, I can see that it says it is empty in the jitdump, but I guess I was expecting something that says the register is explicitly allocated to this node and is available for use.

For example, if I request an internal register, they are explicitly tracked and allocated. They are also listed in the jitdump as attached to the node and are accessible in codegen. I would expect something similar for "fixed" registers (rather than just having them unoccupied, which I am also not seeing a way to check).

Reply:
I don't know of a good way to indicate that on the node without adding extra fields to `GenTree`

Reply:
I don't know of a good way to indicate that on the node without adding extra fields to `GenTree`

Reply:
> For example, if I request an internal register, they are explicitly tracked and allocated.

That's a different thing. There can be multiple internal registers per node so of course they're explicitly tracked and allocated, you need to obtain those registers for a given node. It's also less likely for an internal register to be constrained to a particular physical register so some piece of codegen code could say "hey, I already asked for EBX here so I'll just hardcode it".

> rather than just having them unoccupied, which I am also not seeing a way to check

Reply:
> I don't know of a good way to indicate that on the node without adding extra fields to GenTree

What about tracking it as an internal register, that seems to have all the "necessary" hookups already...

Users could then validate that the desired register is actually allocated to the node, as expected:
```
if (op3Reg != REG_EDI)
{
    assert((node->gtRsvdRegs & RBM_EDI) != 0);
    emit->emitIns_R_R(INS_mov, EA_PTRSIZE, REG_EDI, op3Reg);
}

emit->emitIns_R_R(ins, simdSize, op1Reg, op2Reg);
```

Reply:
> Why do you need to check? You asked for register EDI. Provided that LSRA isn't broken you'll get that register. It's as simple as that. And if LSRA is broken, well, that isn't your problem.

Maybe its just me being a paranoid coder, but I like to be able to actually assert that my assumptions are correct. I don't want to blindly assume that `REG_EDI` is available for use, I want to assert that it is actually available for use before I use it.

This is one of the cases where, if the assumption ever did break, we don't have a good way to catch it outside of tests that fail in possibly obscure manners. Having the ability to assert that it is correct makes it infinitely easier to determine why something is failing because it is clearly called out.

Reply:
That would require special handling in all nodes that require internal registers, as they would have to explicitly exclude that register. Honestly, this has just not been an issue so far so it seems unnecessary. You *could* assert that it is not in `rsMaskVars` (I *think* that's the name), which would at least ensure that no lclVar is in that register.

Reply:
>  I don't want to blindly assume that REG_EDI is available for use, I want to assert that it is actually available for use before I use it.

I don't know, I'd say that it's the norm to assume that another component of the system does its job. I like a few asserts here and there myself, but not because I don't trust other components but because it tends to add more clarity to the code.

This JIT's IR is rather convoluted, with all sorts of corner cases and odd balls, partly a result of it being pushed beyond its original design and partly a result of it trying to be efficient. Asserts help a bit but a better solution would be to have IR validation checks that run after each phase (that modifies IR). There are some but they're not sufficient. I don't know if LSRA has anything like this.

> This is one of the cases where, if the assumption ever did break, we don't have a good way to catch it outside of tests that fail in possibly obscure manners.

Welcome to the optimization and code generation world. There are an infinite number of inputs and all kinds of interactions between various compiler transforms. No matter how many tests you write, no matter how many asserts and checks you have in code you won't be able to catch all bugs. The JIT has even this "stress mode" thing that's intended to catch more bugs by triggering scenarios that are more difficult to trigger using normal tests. Yet bugs slip through and most of the time they happen because someone simply did not think that some particular scenario can happen and thus there are neither tests nor asserts for it.

Reply:
> a better solution would be to have IR validation checks that run after each phase (that modifies IR). There are some but they're not sufficient. I don't know if LSRA has anything like this.

LSRA does have a validation phase (`LinearScan::verifyFinalAllocation()`), though I'm sure it could be improved.